src/main/java/appeng/client/gui/widgets/GuiCraftingTree.java

```diff
@@ -377,21 +426,142 @@ public void draw(int guiMouseX, int guiMouseY) {
+            GL11.glPushMatrix();
+            GL11.glPushAttrib(GL11.GL_ALL_ATTRIB_BITS);
+            final Framebuffer fb = new Framebuffer(fbWidth, fbHeight, true);
+            GL11.glMatrixMode(GL11.GL_PROJECTION);
+            GL11.glLoadIdentity();
+            GL11.glOrtho(0, fbWidth / (float) screenshotZoom, fbHeight / (float) screenshotZoom, 0, 1000, 3000);
+            GL11.glMatrixMode(GL11.GL_MODELVIEW);
+            GL11.glLoadIdentity();
+            GL11.glDisable(GL11.GL_DEPTH_TEST);
+            try {
+                fb.bindFramebuffer(true);
+                for (int xChunk = 0; xChunk < xChunks; xChunk++) {
+                    final int xStart = xChunk * maxGlTexSize;
+                    final int xEnd = Math.min((xChunk + 1) * maxGlTexSize, imgWidth);
+                    final int xChunkSize = xEnd - xStart;
+                    for (int yChunk = 0; yChunk < yChunks; yChunk++) {
+                        final int yStart = yChunk * maxGlTexSize;
+                        final int yEnd = Math.min((yChunk + 1) * maxGlTexSize, imgHeight);
+                        final int yChunkSize = yEnd - yStart;
+
+                        GL11.glPushMatrix();
+                        GL11.glTranslatef(
+                                8 - xStart / (float) screenshotZoom,
+                                8 - yStart / (float) screenshotZoom,
+                                -2000.0f);
+                        GL11.glClear(GL11.GL_COLOR_BUFFER_BIT | GL11.GL_DEPTH_BUFFER_BIT);
+                        draw(-1000, -1000, true);
+                        GL11.glPopMatrix();
+
+                        GL11.glBindTexture(GL11.GL_TEXTURE_2D, fb.framebufferTexture);
+                        GL11.glGetTexImage(
+                                GL11.GL_TEXTURE_2D,
+                                0,
+                                GL12.GL_BGRA,
+                                GL12.GL_UNSIGNED_INT_8_8_8_8_REV,
+                                downloadBuffer);
+
+                        for (int y = 0; y < yChunkSize; y++) {
+                            for (int x = 0; x < xChunkSize; x++) {
+                                outputImg.setRGB(
+                                        x + xStart,
+                                        y + yStart,
+                                        downloadBuffer.get((fbHeight - 1 - y) * fbWidth + x));
+                            }
+                        }
+                    }
+                }
+            } finally {
+                fb.deleteFramebuffer();
+                GL11.glViewport(0, 0, mc.displayWidth, mc.displayHeight);
+            }
+            GL11.glPopAttrib();
+            GL11.glPopMatrix();
+
+            final String date = SCREENSHOT_DATE_FORMAT.format(LocalDateTime.now());
+            String filename = String.format("%s-ae2.png", date);
+            File outFile = new File(screenshotsDir, filename);
+            if (outFile.exists()) {
```

Comment:
I feel like this can be structured in a better way. What do you think?
```java
int idx = 0;
while (outfile.exists()) {
    // Try up to 99 times
    if (idx > 99) {
        throw new FileAlreadyExistsException(filename);
    }

    filename = String.format("%s-ae2-%d.png", date, ++idx);
    outFile = new File(screenshotsDir, filename);
}
```